### ISSUE TITLE
project configuration file

### ISSUE THREAD

**Issue by PavelVozenilek** [c3lang/c3c#8]
The problem: project grows too much to be build by hand. Several compilation targets are needed, too many source files and dependencies between them.

Traditional solution: use makefile. If makefile gets too complicated, try makefile generator. If this doesn't help, why not try something like _bjam_, Turing complete tool for masochists. Or you may try this _XML_ based tool. _XML is the future_, I heard! Big company could hire build engineer.

-----------------

_Could this unsolveable problem be fixed?_

Yes, if the old way of building it object file after object file is dropped.

1. Every project with more than one source file would require mandatory project configuration, a file with fixed name (e.g. `c3-project.cfg`) in project root.
2. This file would contain complete information how to build a project.
3. The compiler will have only one option, which target from the configuration to build. The other parameter would be the project root itself.
    `c3 -debug-with-tests  c:\my_project_root_directory`
    
    `debug-with-tests` is the desired target, fully described in the configuration.
   There will be no other command line options available in the compiler.
4. The compiler would then read project configuration, extract what to build and how, and will do it.

-----------

_How would the project configuration look like?_

Like an INI file. Not like makefile or other abominations. There would be no smartness inside, no trickery, no shows how clever one can get.

Example:

```
[global-settings]
enforce-parenthesis-around-binary-operators = *true | false
implicit-paddings-allowed-in-structs = true | *false
allow-utf8-characters-in-soure = none | *in-strings-and_comments | anywhere
...

// every single compilation target described here, 
// completely and independently to other target
[debug-with-tests]
defines = DEBUG, TESTS
optimization-option = *O0 | O1 | O2
libraries-path=...
...

[debug-no-tests]
defines = DEBUG
optimization-option = *O0 | O1 | O2
libraries-path=...
...

[release-with-tests]
....

[release-full-speed]
...
```

--------------

_Wouldn't it mean lot of duplication?_

Yes, and that's intentional.

If you use the same file path on 20 different places, so be it. Copy it 20x, and if you need to change it, do it in all 20 places.

There should be no possible doubt how project is being build. Any smartness, shared settings, settings computed on the flight, clever use of tool features is strictly forbidden.

--------------

_Could such configuration be made foolproof?_

Yes and this is the most important part of this post. The compiler would know exactly what the configuration has to contain and will complain if something goes wrong.

1. The compiler will check that all items expected in configuration are present, in the correct order, without typos. If e.g. global settings `implicit-paddings-allowed-in-structs` is missing, it would be error. Error message should say what is expected.
2. The compiler will refuse all items it does not understand (typos, previous version settings). No garbage would be allowed in the configuration.
3. All available options for a settings need to be present in configuration file:
    ```
    implicit-paddings-allowed-in-structs = true | *false
    optimization-option = *O0 | O1 | O2
    ```

    This in order to make it absolutely clear what are the available options, to avoid guessing games. Exactly one of these options has to be selected, e.g. by `*`. No typo or wrong order among the options would be allowed.

4. Options which are not finite (e.g. list of libraries) would be verified as much as possible. Paths would be checked for validity. No effort to make it foolproof should be spared.

5. Problems during building should result in complete error information. No effort should be spared here either.

--------------
_Wouldn't it be lot of work to implement?_
Yes, it would.

_Is there some way to run other build systems (e.g for the C part of the project)?_
Yes, there could be some command to run external process. It would be at user's full risk.

_What are some other advantages of this approach?_
1. It does not rely on cryptic and fragile compiler command line. There's exactly one place where all nicely named build options are stored. Unlike makefiles  these options are easy to edit.
2. Creating new build target is easy. Just copy the most close section and modify it. If you do not know what these options mean, bad luck. Deleting target is equally easy.
3. You see it everything in one place, in expected order, and with all possible options. This is something what no other system offers.
4. Typical mistake - "Oops, I forgot to specify this" - will be very limited.

_Jai uses its language to create build the targets._
It is not needed.

_Should the code have access to used build options during compile time?_
Yes. An artificial example (not recommended to be used in practice): depending on optimization for size or speed, during compile time different code path could be selected by compile time if.

_What if I make mistake in the configuration?_
Your problem. At least, you have it all in one place, this makes it easier to spot bugs.

---------------

Hypothetical, very advanced IDE, could show what are effects of configuration changes within the project. E.g. which code is used because of this option, where that option makes a difference.

**Comment by PavelVozenilek** [comment]
More useful example for this feature:

> _Should the code have access to used build options during compile time?_
> Yes. An artificial example (not recommended to be used in practice): depending on optimization for size or speed, during compile time different code path could be selected by compile time if.

Piece of code or a library could, during compile time, check configuration settings and complain if they contain wrong combination, or if there's missing some define. This is, incompletely and clumsily, possible in C via macros.

For example, code may require full optimization on, when it compiles in tests specifically used to guard against performance regressions.

**Comment by lerno** [comment]
@PavelVozenilek This parallels what's written here in many regards: https://c3lang.github.io/c3docs/project/ can you summarize the differences/improvements you suggest?

**Comment by PavelVozenilek** [comment]
> https://c3lang.github.io/c3docs/project/

I haven't seen this yet - I am first going through simpler parts of the docs, making notes as it goes.

The basic idea looks quite similar, though I'd avoid un-needed things like _authors_. I would also remove all dependencies on environment. Use of environment could cause headaches, it is something not clearly visible and fixed in letters.  IMHO environment is always misfeature. `$HOME` is "Unixism", another reason to avoid it.


IMO the configuration should know about:
1. Project root (where the config file sits), known as "`.`" (and using "`/`" path separator,  which is understood on Windows). The ideal project would have everything inside the project tree and no need to care about the rest of filesystem.

2. Where relative path from project root is not feasible, one should use absolute _hardcoded_ paths valid in the current OS. If someone wants to compile the same project on another OS, he should clone the target section a modify these absolute paths. This way mistake/confusion in some target won't affect the rest.


Major differences/improvements:
* In my proposal making new target is something cheap. Copy existing section, make few changes, and that's it. This isolates potential problems. If I screw up completely in one place, the rest should still work.
* I am very strict with contents of configuration file. There should be no freedom to deviate from the expected form. If something could go wrong, it will. Not allowing to go wrong is the solution.
* According to me configuration file would the _only_ way to compile non-trivial (> 1 source file) project.  Single source could still use traditional command line options or whatever else.

**Comment by lerno** [comment]
1. As long as you are *anywhere* inside of the project structure, the compiler will walk up through the tree to find `project.toml`. See for details of the mandatory structure: https://c3lang.github.io/c3docs/firstproject/

All paths are relative to that that file, which is placed in the project root. And yes, if platform specific code is needed then another target is created with those settings.

- Adding a target works very much like the copy/paste you suggest.
- The project structure is created from the compiler and is the required form for compiling anything other than single files.

**Comment by PavelVozenilek** [comment]
OK. The main (and possibly the only) difference is, that I do not expect the project to have any predefined internal structure.


I hate tools making decisions how project tree should look like. Go language does this - yet another reason for me to avoid that dreck. I have very strong opinions about minimalism, while these tools try to please everyone generating impressive amount of garbage.

**Comment by lerno** [comment]
Some parts of the structure is necessary to be similar in order to have fewer unnecessary knobs and switches.

**Comment by PavelVozenilek** [comment]
About https://github.com/c3lang/c3c/issues/8#issuecomment-521918547:

I do not see it as necessary. My ideal is (mostly) flat structure, with numerically labeled files and also textual doc files placed inbetween.

Here are parts of generated project tree I do not like (and why):
* `build/`: the whole build system could be one configuration file.  If whole project is built as single source, no intermediate build artifacts are generated.
* `lib/`: I suppose it is for 3rd party libraries. This arrangement places them away from locations where they would fit best. What is the value of having all foreign libraries in one place? To be able to delete them all at once? `lib/` is victory of organization structure (we versus the others) over utility.
* `src/`: it moves sources form natural place (in the root) one level away. Unnecessary complication.
* `tests/`: I proposed system of easy tests with no noise, like the need for separate directory.


OTOH, I sometimes find a need to have `temp/` directory in project root, for dirty things. E.g. I may want to write down text file with currently running test file name and file, when some unknown test causes crash which cannot be caught with the IDE.


In my opinion any generated project structure should be optional feature, not something forced down one's throat.

**Comment by lerno** [comment]
Build artifacts are a fact when first transpiling to C. The resulting executable should be placed in a predictable place, not where the compilation was invoked.

But if you want to have everything in a single blob that's possible. Just change the project settings from the defaults. E.g. `libs = ["./**"]` or `libs = ["."]` if you don't want recursive inclusion.

**Comment by lerno** [comment]
Please file this on c3docs.
